konnectivity-client/pkg/client/metrics/metrics.go

```diff
@@ -0,0 +1,139 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	sharedmetrics "sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/shared/metrics"
+	"sigs.k8s.io/apiserver-network-proxy/konnectivity-client/proto/client"
+)
+
+var (
+	// Metrics provides access to all client metrics. The client
+	// application is responsible for registering (via Metrics.RegisterMetrics).
+	Metrics = newMetrics()
+)
+
+// ClientMetrics includes all the metrics of the konnectivity-client.
+type ClientMetrics struct {
+	namespace    string
+	subsystem    string
+	registerOnce sync.Once
+	// registered is an atomic bool represented as a 0 or 1, and set to 1 when the metrics are registered.
+	registered uint32
+
+	tConnections       *prometheus.GaugeVec
+	streamEvents       *prometheus.CounterVec
+	streamEventsErrors *prometheus.CounterVec
+}
+
+func newMetrics() *ClientMetrics {
+	return &ClientMetrics{
+		registered:         0,
+		tConnections:       nil,
+		streamEvents:       nil,
+		streamEventsErrors: nil,
+	}
+}
+
+func (t *ClientMetrics) isRegistered() bool {
+	return atomic.LoadUint32(&t.registered) != 0
+}
+
+func (c *ClientMetrics) populateMetrics(namespace, subsystem string) {
```

Comment:
We can use whatever string makes sense, but I would not actually recommend parameterizing namespace and subsystem because it breaks our ability to run static analysis.

Reply:
In this case, I'll use the same namespace as server and agent ('konnectivity_network_proxy') and add subsystem 'client'.